[BUG] The concurrent writer throws a class casting error when enabling AQE.

**Describe the bug**
@HaoYang670 found this bug, thanks.
The concurrent writer throws a class casting error when enabling AQE.

Error is:
```
Caused by: java.lang.ClassCastException: org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec cannot be cast to com.nvidia.spark.rapids.GpuExec
```

Related [code](https://github.com/NVIDIA/spark-rapids/blob/branch-22.12/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatWriter.scala#L226) is:

```
        if (concurrentWritersEnabled) {
          (empty2NullPlan.executeColumnar(),
              Some(GpuConcurrentOutputWriterSpec(maxWriters, empty2NullPlan.asInstanceOf[GpuExec],
                orderingExpr)))
        } else {
```

If AQE is enabled, `empty2NullPlan` is actually a `AdaptiveSparkPlanExec`.

**Steps/Code to reproduce bug**
```
@ignore_order
@allow_non_gpu(any = True)
def test_my_test(spark_tmp_path):
    data_path = spark_tmp_path + '/PARQUET_DATA'
    assert_gpu_and_cpu_writes_are_equal_collect(
        lambda spark, path: get_25_partitions_df(spark)  # df has 25 partitions for (c1, c2)
            .repartition(2)
            .write.mode("overwrite").partitionBy('c1', 'c2').parquet(path),
        lambda spark, path: spark.read.parquet(path),
        data_path,
        copy_and_update(
            # 10 < 25, will fall back to single writer
            {"spark.sql.maxConcurrentOutputFileWriters": 10},
            {"spark.sql.adaptive.enabled": "true"},    #  enable AQE
        ))
```

--- a/integration_tests/src/main/python/parquet_write_test.py
+++ b/integration_tests/src/main/python/parquet_write_test.py
@@ -524,8 +524,8 @@ def test_concurrent_writer(spark_tmp_path):
 @ignore_order
 @pytest.mark.skipif(is_before_spark_320(), reason="is only supported in Spark 320+")
 @allow_non_gpu(any=True)
-@pytest.mark.parametrize('enable_aqe', ["true", "false"])
-def test_fallback_to_single_writer_from_concurrent_writer(spark_tmp_path, enable_aqe):
+@pytest.mark.parametrize('aqe_enabled', [True, False])
+def test_fallback_to_single_writer_from_concurrent_writer(spark_tmp_path, aqe_enabled):
     data_path = spark_tmp_path + '/PARQUET_DATA'
     assert_gpu_and_cpu_writes_are_equal_collect(
         lambda spark, path: get_25_partitions_df(spark)  # df has 25 partitions for (c1, c2)
@@ -537,7 +537,7 @@ def test_fallback_to_single_writer_from_concurrent_writer(spark_tmp_path, enable
             # 10 < 25, will fall back to single writer
             {"spark.sql.maxConcurrentOutputFileWriters": 10},
             {"spark.rapids.sql.concurrentWriterPartitionFlushSize": 64 * 1024 * 1024},
-            {"spark.sql.adaptive.enabled": enable_aqe},
+            {"spark.sql.adaptive.enabled": aqe_enabled},
         ))
 
 

--- a/integration_tests/src/main/python/parquet_write_test.py
+++ b/integration_tests/src/main/python/parquet_write_test.py
@@ -523,7 +523,9 @@ def test_concurrent_writer(spark_tmp_path):
 
 @ignore_order
 @pytest.mark.skipif(is_before_spark_320(), reason="is only supported in Spark 320+")
-def test_fallback_to_single_writer_from_concurrent_writer(spark_tmp_path):
+@allow_non_gpu(any=True)
+@pytest.mark.parametrize('enable_aqe', ["true", "false"])
+def test_fallback_to_single_writer_from_concurrent_writer(spark_tmp_path, enable_aqe):
     data_path = spark_tmp_path + '/PARQUET_DATA'
     assert_gpu_and_cpu_writes_are_equal_collect(
         lambda spark, path: get_25_partitions_df(spark)  # df has 25 partitions for (c1, c2)
@@ -534,7 +536,8 @@ def test_fallback_to_single_writer_from_concurrent_writer(spark_tmp_path):
         copy_and_update(
             # 10 < 25, will fall back to single writer
             {"spark.sql.maxConcurrentOutputFileWriters": 10},
-            {"spark.rapids.sql.concurrentWriterPartitionFlushSize": 64 * 1024 * 1024}
+            {"spark.rapids.sql.concurrentWriterPartitionFlushSize": 64 * 1024 * 1024},
+            {"spark.sql.adaptive.enabled": enable_aqe},
         ))
 
 

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala
@@ -69,10 +69,14 @@ class GpuSortMeta(
 
 object GpuSortExec {
   def targetSize(sqlConf: SQLConf): Long = {
+    val batchSize = RapidsConf.GPU_BATCH_SIZE_BYTES.get(sqlConf)
+    targetSize(batchSize)
+  }
+
+  def targetSize(batchSize: Long): Long = {
     // To avoid divide by zero errors, underflow and overflow issues in tests
     // that want the targetSize to be 0, we set it to something more reasonable
-    val targetSize = RapidsConf.GPU_BATCH_SIZE_BYTES.get(sqlConf)
-    math.max(16 * 1024, targetSize)
+    math.max(16 * 1024, batchSize)
   }
 }
 

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala
@@ -779,7 +779,7 @@ class GpuDynamicPartitionDataConcurrentWriter(
    */
   private def getSorted(iterator: Iterator[ColumnarBatch]): GpuOutOfCoreSortIterator = {
     val gpuSortOrder: Seq[SortOrder] = spec.sortOrder
-    val output: Seq[Attribute] = spec.exec.output
+    val output: Seq[Attribute] = spec.output
     val sorter = new GpuSorter(gpuSortOrder, output)
     val cpuOrd = new LazilyGeneratedOrdering(sorter.cpuOrdering)
 
@@ -795,7 +795,8 @@ class GpuDynamicPartitionDataConcurrentWriter(
 
       override def semaphoreWaitTime: GpuMetric = NoopMetric
     }
-    val targetSize = GpuSortExec.targetSize(spec.exec.conf)
+
+    val targetSize = GpuSortExec.targetSize(spec.batchSize)
     // out of core sort the entire iterator
     GpuOutOfCoreSortIterator(iterator, sorter, cpuOrd, targetSize,
       opTime, sortTime, outputBatch, outputRows,

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatWriter.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatWriter.scala
@@ -81,8 +81,8 @@ object GpuFileFormatWriter extends Logging {
   }
 
   /** Describes how concurrent output writers should be executed. */
-  case class GpuConcurrentOutputWriterSpec(maxWriters: Int, exec: GpuExec,
-      sortOrder: Seq[SortOrder])
+  case class GpuConcurrentOutputWriterSpec(maxWriters: Int, output: Seq[Attribute],
+      batchSize: Long, sortOrder: Seq[SortOrder])
 
   /**
    * Basic work flow of this command is:
@@ -222,8 +222,9 @@ object GpuFileFormatWriter extends Logging {
           requiredOrdering
               .map(attr => SortOrder(attr, Ascending)), finalOutputSpec.outputColumns)
         if (concurrentWritersEnabled) {
+          val batchSize = RapidsConf.GPU_BATCH_SIZE_BYTES.get(sparkSession.sessionState.conf)
           (empty2NullPlan.executeColumnar(),
-              Some(GpuConcurrentOutputWriterSpec(maxWriters, empty2NullPlan.asInstanceOf[GpuExec],
+              Some(GpuConcurrentOutputWriterSpec(maxWriters, empty2NullPlan.output, batchSize,
                 orderingExpr)))
         } else {
           val sortType = if (useStableSort) {

